shared/protocol/src/main/java/io/pravega/shared/protocol/netty/CommandEncoder.java

```diff
@@ -72,13 +81,102 @@
+
+        /**
+         * Record the given Append to Session by tracking the last event number and number of events to write.
+         */
+        private void record(Append append) {
+            lastEventNumber = append.getEventNumber();
+            eventCount += append.getEventCount();
+        }
+
+        /**
+         * Check if session is empty or not.
+         * @return true if there are any pending events; false otherwise.
+         */
+        private boolean isFree() {
+            return eventCount == 0;
+        }
+
+
+        /**
+         * queue the Append data to Session' list.
+         * @param data  data bytes.
+         * @param out   Network channel buffer.
+         */
+        private void write(ByteBuf data, ByteBuf out) {
+            pendingWrites.putIfAbsent(id, this);
+            if (data.readableBytes() > 0) {
+                pendingBytes += data.readableBytes();
+                pendingList.add(data);
+            }
+            conditionalFlush(out);
+        }
+
+        /**
+         * check the overflow condition and flush if required.
+         * @param out   Network channel buffer.
+         */
+        private void conditionalFlush(ByteBuf out) {
+            if ((pendingBytes > MAX_DATA_SIZE) || (eventCount > MAX_EVENTS)) {
+                breakCurrentAppend(out);
+                flush(out);
+            }
+        }
+
+       /**
+        * Write/flush session's data to network channel.
+        * @param out   Network channel buffer.
+        */
+       private void flush(ByteBuf out) {
+            if (!isFree()) {
+                pendingWrites.remove(id);
+                if (pendingBytes > 0) {
+                    writeMessage(new AppendBlock(id), pendingBytes, out);
+                    for (Iterator<ByteBuf> iterator = pendingList.iterator(); iterator.hasNext(); ) {
+                        out.writeBytes(iterator.next());
+                        iterator.remove();
+                    }
+                }
+                flush(pendingBytes, null, out);
+                pendingBytes = 0;
+            }
+        }
+
+        /**
+         * Write/flush given data to network channel.
+         * by writing the Append block End containing the given/input data and indicating the size of previous
+         * append block data.
+         * @param out   Network channel buffer.
+         */
+        private void flush(int sizOfWholeEvents, ByteBuf data, ByteBuf out) {
+            writeMessage(new AppendBlockEnd(id, sizOfWholeEvents, data, eventCount, lastEventNumber, requestId), out);
+            eventCount = 0;
+        }
+    }
+
+    /**
+     * Write/flush Multi session's buffered data to network channel buffer
+     * by iterating over all the pending sessions .
+     * @param out   Network Channel Buffer.
+     */
+    private void flushAll(ByteBuf out) {
+        if (!pendingWrites.isEmpty()) {
+            ArrayList<Session> sessions = new ArrayList<>(pendingWrites.values());
```

Comment:
Why do you make a copy and then iterate through the copy? There's no benefit to this. Did you intend to clear the `pendingWrites` map?

Reply:
yes; the flush removes them PendingWrites;   hence ; pending sessions are collected first and for each session flush is invoked. flush removes itself from the pendingWrites Map.